cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -2083,14 +2083,18 @@ private class PathNodeSink extends PathNode, TPathNodeSink {
  * a callable is recorded by `cc`.
  */
 private predicate pathStep(PathNodeMid mid, Node node, CallContext cc, SummaryCtx sc, AccessPath ap) {
-  exists(LocalCallContext localCC, AccessPath ap0, Node midnode, Configuration conf |
+  exists(
+    LocalCallContext localCC, AccessPath ap0, Node midnode, Configuration conf,
+    DataFlowCallable call
+  |
     midnode = mid.getNode() and
     conf = mid.getConfiguration() and
     cc = mid.getCallContext() and
     sc = mid.getSummaryCtx() and
-    localCC = getLocalCallContext(cc, midnode.getEnclosingCallable()) and
+    call = midnode.getEnclosingCallable() and
```

Comment:
This is not behaviour-preserving.  It looks like the root of the problem is that `getLocalCallContext` has been inlined - could you try adding `pragma[nomagic]` to that predicate instead?

Reply:
(Also, minor aside, a callable should not be named "call" - as it isn't a `Call`).

Reply:
Adding `nomagic` to `getLocalCallContext ` gives the same bad join:
```
Tuple counts for DataFlowImplLocal::pathStep#fffff#cur_delta:
                      83       ~0%        {6} r1 = SCAN DataFlowImplLocal::PathNodeMid#class#ffffff#prev_delta AS I OUTPUT I.<2>, I.<0>, I.<1>, I.<3>, I.<4>, I.<5>
                      51077304 ~4%        {7} r2 = JOIN r1 WITH DataFlowImplCommon::ImplCommon::CallContext::relevantFor_dispred#ff AS R ON FIRST 1 OUTPUT r1.<2>, R.<1>, r1.<1>, r1.<0>, r1.<3>, r1.<4>, r1.<5>
                      83       ~0%        {7} r3 = JOIN r2 WITH DataFlowUtil::Node::getFunction_dispred#ff AS R ON FIRST 2 OUTPUT r2.<0>, true, r2.<6>, r2.<2>, r2.<3>, r2.<4>, r2.<5>
                      0        ~0%        {5} r4 = JOIN r3 WITH DataFlowImplLocal::localFlowBigStep#fffff_0231#join_rhs AS R ON FIRST 3 OUTPUT r3.<3>, r3.<4>, r3.<5>, r3.<6>, R.<3>
```

And you're right about the `call` name. I will keep that in mind for future commits.

Reply:
I incidentally also ran into this, here is my proposed fix: https://github.com/Semmle/ql/pull/2822/files#diff-16640961876041be2d06fbe5466f9c07R2221